File: SwrveSDK/SDK/Messaging/SwrveEmbeddedMessage.m
#import "SwrveEmbeddedMessage.h"
#import "SwrveMessageController.h"

@interface SwrveEmbeddedMessage()

@property (nonatomic, weak) SwrveMessageController* controller;

@end

@implementation SwrveEmbeddedMessage

@synthesize controller, campaign, data, type, buttons;

-(id)initWithDictionary:(NSDictionary *)json forCampaign:(SwrveEmbeddedCampaign *)_campaign forController:(SwrveMessageController*)_controller {
    if (self = [super init]) {
        self.campaign     = _campaign;
        self.controller   = _controller;
        self.messageID    = [json objectForKey:@"id"];
        if ([json objectForKey:@"priority"]) {
            self.priority = [json objectForKey:@"priority"];
        } else {
            self.priority = [NSNumber numberWithInt:9999];
        }
        
        if ([json objectForKey:@"buttons"] && [json objectForKey:@"buttons"] != [NSNull null]) {
            self.buttons = [json mutableArrayValueForKey:@"buttons"];
        }
        
        if ([json objectForKey:@"type"] && [json objectForKey:@"type"] != [NSNull null]) {
            self.type = [self retrieveType:[json objectForKey:@"type"]];
        }
        
        if ([json objectForKey:@"data"] && [json objectForKey:@"data"] != [NSNull null]) {
            self.data = [json objectForKey:@"data"];
        }
        
        if ([json objectForKey:@"name"] && [json objectForKey:@"name"] != [NSNull null]) {
            self.name = [json objectForKey:@"name"];
        }
        
        if ([json objectForKey:@"message_center_details"] && [json objectForKey:@"message_center_details"] != [NSNull null]) {
            self.messageCenterDetails = [[SwrveMessageCenterDetails alloc] initWithJSON:[json objectForKey:@"message_center_details"]];
        }
        
        if ([json objectForKey:@"control"] && [json objectForKey:@"control"] != [NSNull null]) {
            self.control = [[json objectForKey:@"control"] boolValue];
        }
    }
    return self;
}

-(SwrveEmbeddedDataType) retrieveType:(NSString *) typeString {
    if([typeString isEqualToString:@"json"]) {
        return kSwrveEmbeddedDataTypeJson;
    }
        
    return kSwrveEmbeddedDataTypeOther;
}

@end

File: SwrveSDKSwift/SwrveEmbeddedMessage.swift
import Foundation
import UIKit

#if canImport(SwrveSDK)
    import SwrveSDK
#endif

#if canImport(SwrveSDKCommon)
    import SwrveSDKCommon
#endif

@objc public enum SwrveEmbeddedDataType: Int {
    case json
    case other
}

@objc public class SwrveEmbeddedMessage: SwrveBaseMessage {

    @objc public var data: String?
    @objc public var type: SwrveEmbeddedDataType = .other
    @objc public var buttons: [String] = []

    @objc public static func fromDictionary(_ json: [String: Any], andCampaign campaign: SwrveCampaign) -> SwrveEmbeddedMessage {
        let message = SwrveEmbeddedMessage()
        message.campaign = campaign

        guard let id = json["id"] as? NSNumber else {
            return message
        }

        message.messageID = id
        message.priority = json["priority"] as? NSNumber ?? NSNumber(value: 9999)
        message.buttons = json["buttons"] as? [String] ?? []

        let typeString = json["type"] as? String
        message.type = typeString == "json" ? .json : .other

        message.data = json["data"] as? String
        message.name = json["name"] as? String

        if let messageCenterDetailsJSON = json["message_center_details"] as? [String: Any] {
            message.messageCenterDetails = SwrveMessageCenterDetails(withJSON: messageCenterDetailsJSON)
        }

        if let control = json["control"] as? Bool {
            message.control = control
        }

        return message
    }

}
